### ISSUE TITLE
max_insert_threads and materialized views performance degradation

### ISSUE THREAD

**Issue by filimonov** [ClickHouse/ClickHouse#37900]
test
```sql
DROP TABLE IF EXISTS t;
DROP TABLE IF EXISTS t_mv;
create table t (a UInt64) Engine = Null;
create materialized view t_mv Engine = Null AS select now() as ts, max(a) from t group by ts;
insert into t select * from numbers_mt(3000000000) settings max_threads = 16, max_insert_threads=16;
```

results:
```
21.8:  
0 rows in set. Elapsed: 3.841 sec. Processed 3.00 billion rows, 24.00 GB (781.01 million rows/s., 6.25 GB/s.)

21.10: 
0 rows in set. Elapsed: 3.844 sec. Processed 6.00 billion rows, 48.00 GB (1.56 billion rows/s., 12.49 GB/s.)

21.11:
0 rows in set. Elapsed: 17.717 sec. Processed 5.98 billion rows, 47.86 GB (337.70 million rows/s., 2.70 GB/s.)

21.12
0 rows in set. Elapsed: 18.569 sec. Processed 6.00 billion rows, 48.00 GB (323.12 million rows/s., 2.58 GB/s.)

22.1
0 rows in set. Elapsed: 19.124 sec. Processed 6.00 billion rows, 48.00 GB (313.74 million rows/s., 2.51 GB/s.)

22.2
0 rows in set. Elapsed: 19.939 sec. Processed 6.00 billion rows, 48.00 GB (300.92 million rows/s., 2.41 GB/s.)

22.3
0 rows in set. Elapsed: 18.727 sec. Processed 6.00 billion rows, 48.00 GB (320.39 million rows/s., 2.56 GB/s.)

22.4
0 rows in set. Elapsed: 19.698 sec. Processed 6.00 billion rows, 48.00 GB (304.60 million rows/s., 2.44 GB/s.)

22.5
0 rows in set. Elapsed: 22.723 sec. Processed 6.00 billion rows, 48.00 GB (264.05 million rows/s., 2.11 GB/s.)
```

SIde note: pay attention to processed rows and progress bar behavior (minor cosmetic issue).

**Comment by hodgesrm** [comment]
I've encountered this on a large data load. Performance degradation was about 90% compared to 21.8 LTS.

**Comment by anotherbugmaster** [comment]
I have the same issue. It seems that MV is the culprit here because `max_insert_threads` setting scale performance correctly when MV is not being used

**Comment by hodgesrm** [comment]
Hello Nicolai! Is there anything we can do to help with resolving this problem?

**Comment by azat** [comment]
I've started looking into this before, first I found that clang-11 generates slightly better code for `max`.
Then I've generated flamegraphs but it does not shows that significant difference.
Then I tried to compile upstream with clang-11, but failed (clang fails with some ICE).

That I've looked at views and `max_insert_threads` was simply broken, this should be fixed by #38731
Also I noticed that `optimize_trivial_insert_select=1` can break parallel views processing again, this had been fixed too.
